rts/src/main/java/eta/runtime/io/ByteArray.java

```diff
@@ -41,39 +41,81 @@ private ByteArray(long bufferAddress, int size) {
-                                    , ByteArray destArray, int destOffset
-                                    , int n) {
-        ByteBuffer src  = MemoryManager.getBoundedBuffer(srcArray.bufferAddress);
-        ByteBuffer dest = MemoryManager.getBoundedBuffer(destArray.bufferAddress);
-        src.position(src.position() + srcOffset);
-        src.limit(src.position() + n);
-        dest.position(dest.position() + destOffset);
-        dest.put(src);
+    public ByteBuffer getBuffer() {
+        return MemoryManager.getBoundedBuffer(bufferAddress);
+    }
+
+    public ByteBuffer getBuffer(int offset, int n) {
+        return MemoryManager.getBoundedBuffer(bufferAddress, offset, n);
+    }
+
+    public static ByteBuffer copyAddrToByteArray( long srcAddress, ByteArray destArray
+                                                , int offset, int n) {
+        return MemoryManager.copy(srcAddress,offset,
+                                  destArray.getBuffer(),offset,n);
+    }
+
+    public static ByteBuffer copyByteArrayToAddr( ByteArray srcArray, int offset
+                                                , long destAddress, int n) {
+        return MemoryManager.copy(srcArray.getBuffer(),offset,
+                                  destAddress,offset,n);
+    }
+
+    public static ByteBuffer copyByteArray( ByteArray srcArray, int srcOffset
+                                          , ByteArray destArray, int destOffset
+                                          , int n) {
+        return MemoryManager.copyByteBuffer(srcArray.getBuffer(),srcOffset,
+                                  destArray.getBuffer(),destOffset,n);
+    }
+
+    public static ByteBuffer copyByteArray( ByteArray srcArray, int srcOffset
+                                          , ByteBuffer dest, int destOffset
+                                          , int n) {
+        return MemoryManager.copyByteBuffer(srcArray.getBuffer(),srcOffset,
+                                  dest,destOffset,n);
+    }
+
+    public static ByteBuffer copyByteArray( ByteBuffer src, int srcOffset
+                                           , ByteArray destArray, int destOffset
+                                           , int n) {
+        return MemoryManager.copyByteBuffer(src,srcOffset,
+                                  destArray.getBuffer(),destOffset,n);
+    }
+
+    public static ByteBuffer copyByteArray(ByteArray srcArray, ByteArray destArray, int size) {
+        return copyByteArray(srcArray, 0, destArray, 0, size);
+    }
+
+    public static ByteBuffer copyByteArray(ByteBuffer srcBuffer, ByteArray destArray, int size) {
+        return copyByteArray(srcBuffer, 0, destArray,0, size);
+    }
+
+    public static ByteBuffer copyByteArray(ByteArray srcArray, ByteBuffer destBuffer, int size) {
+        return copyByteArray(srcArray, 0, destBuffer, 0, size);
+    }
+
+    public static int compare(ByteArray b1, ByteArray b2, int n) {
```

Comment:
From this line (103) and below, all the uses of `duplicate()` are unnecessary since the ByteBuffers are created as intermediates in all the cases and won't leak outside these methods.